VariantShredding.md

```diff
@@ -25,290 +25,318 @@
-Similarly, for the query `select * from tbl where variant_get(variant_col, ‘$.id’, ‘integer’) = 123`, the scan could first decode the shredded `id` column, and only fetch/decode the full Variant value for rows that pass the filter.
+Variant metadata is stored in the top-level Variant group in a binary `metadata` column regardless of whether the Variant value is shredded.
 
-# Parquet Example
+All `value` columns within the Variant must use the same `metadata`.
+All field names of a Variant, whether shredded or not, must be present in the metadata.
 
-Consider the following Parquet schema together with how Variant values might be mapped to it.
-Notice that we represent each shredded field in `object` as a group of two fields, `typed_value` and `variant_value`.
-We extract all homogenous data items of a certain path into `typed_value`, and set aside incompatible data items in `variant_value`.
-Intuitively, incompatibilities within the same path may occur because we store the shredding schema per Parquet file, and each file can contain several row groups.
-Selecting a type for each field that is acceptable for all rows would be impractical because it would require buffering the contents of an entire file before writing.
+## Value Shredding
 
-Typically, the expectation is that `variant_value` exists at every level as an option, along with one of `object`, `array` or `typed_value`.
-If the actual Variant value contains a type that does not match the provided schema, it is stored in `variant_value`.
-An `variant_value` may also be populated if an object can be partially represented: any fields that are present in the schema must be written to those fields, and any missing fields are written to `variant_value`.
-
-The `metadata` column is unchanged from its unshredded representation, and may be referenced in `variant_value` fields in the shredded data.
+Variant values are stored in Parquet fields named `value`.
+Each `value` field may have an associated shredded field named `typed_value` that stores the value when it matches a specific type.
+When `typed_value` is present, readers **must** reconstruct shredded values according to this specification.
 
+For example, a Variant field, `measurement` may be shredded as long values by adding `typed_value` with type `int64`:
 ```
-optional group variant_col {
- required binary metadata;
- optional binary variant_value;
- optional group object {
-  optional group a {
-   optional binary variant_value;
-   optional int64 typed_value;
-  }
-  optional group b {
-   optional binary variant_value;
-   optional group object {
-    optional group c {
-      optional binary variant_value;
-      optional binary typed_value (STRING);
-    }
-   }
-  }
- }
+required group measurement (VARIANT) {
+  required binary metadata;
```

Comment:
I'm not sure if this is already mentioned anywhere, but there's no requirement of ordering for these fields in the parquet schema, right? I wouldn't expect there to be, but the MapType in https://github.com/apache/parquet-format/blob/master/LogicalTypes.md is a bit ambiguous (`key and value can be identified by their position in case of misnaming`), so it might be good to explicitly mention that the metadata doesn't need to come first in the schema.

Reply:
I think that's a good point. I'll note that the names must match and while the order should match, it should not be relied on.